### PR TITLE
Add GreedySoup mini-ensemble algorithm

### PR DESCRIPTION
**Pull request by sxjscience**
*Description of changes:*

Implement the GreedySoup algorithm proposed in https://arxiv.org/pdf/2203.05482.pdf

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: text/src/autogluon/text/automm/predictor.py [hunk 1]
```diff
@@ -566,21 +566,65 @@ def _fit(
...
+                                              key=lambda ele: ele[1])
+                best_k_models_path = [ele[0] for ele in best_k_models_sorted]
+            elif minmax_mode == 'max':
+                best_k_models_sorted = sorted(list(checkpoint_callback.best_k_models.items()),
+                                              key=lambda ele: ele[1],
+                                              reverse=True)
+                best_k_models_path = [ele[0] for ele in best_k_models_sorted]
+            else:
+                raise ValueError(f'Unsupported minmax_mode={minmax_mode}')
+            print(best_k_models_sorted)
```
Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:581]: remove print statement

File: text/src/autogluon/text/automm/predictor.py [hunk 2]
```diff
@@ -566,21 +566,65 @@ def _fit(
...
+            # Obtain the best K models sorted from the best performance to the worst performance
+            if minmax_mode == 'min':
+                best_k_models_sorted = sorted(list(checkpoint_callback.best_k_models.items()),
+                                              key=lambda ele: ele[1])
+                best_k_models_path = [ele[0] for ele in best_k_models_sorted]
+            elif minmax_mode == 'max':
+                best_k_models_sorted = sorted(list(checkpoint_callback.best_k_models.items()),
+                                              key=lambda ele: ele[1],
+                                              reverse=True)
+                best_k_models_path = [ele[0] for ele in best_k_models_sorted]
```
Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:578]: This kind of logic is pretty prone to bugs, any thoughts on standardizing to always be lower=better regardless of metric to simplify downstream code? This is done in Tabular and vastly simplified a lot of logic + reduced bugs

Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:578]: This wasn't present before. Was it bugged previously? Why is this code needed now? Is it because we care about the order of the top K since we aren't simply averaging?

File: text/src/autogluon/text/automm/predictor.py [hunk 3]
```diff
@@ -566,21 +566,65 @@ def _fit(
...
+                    model = self._load_state_dict(
+                        model=model,
+                        state_dict=cand_avg_state_dict,
+                    )
+                    self._model = model
+                    cand_performance = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                    if minmax_mode == 'min' and cand_performance < best_performance or \
+                       minmax_mode == 'max' and cand_performance > best_performance:
+                        ingredients.append(all_state_dicts[i])
+                        best_performance = cand_performance
```
Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:612]: Consider moving to a separate method

File: text/src/autogluon/text/automm/predictor.py [hunk 4]
```diff
@@ -566,21 +566,65 @@ def _fit(
...
+                        ckpt_template=ckpt_template
+                    )
+                    model = self._load_state_dict(
+                        model=model,
+                        state_dict=cand_avg_state_dict,
+                    )
+                    self._model = model
+                    cand_performance = self.evaluate(val_df, [validation_metric_name])[validation_metric_name]
+                    if minmax_mode == 'min' and cand_performance < best_performance or \
+                       minmax_mode == 'max' and cand_performance > best_performance:
```
Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:610]: ditto regarding considering force lower=better to simplify code

File: text/src/autogluon/text/automm/predictor.py [hunk 5]
```diff
@@ -566,21 +566,65 @@ def _fit(
...
+                        ingredients.append(all_state_dicts[i])
+                        best_performance = cand_performance
+            else:
+                ingredients = all_state_dicts
```
Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:614]: raise helpful ValueError if unknown `config.optimization.top_k_average_type`. Should only accept: `"union_soup"` and `"greedy_soup"`. This error should tell the user what they specified and what the valid values are.

Comment (Innixma) [inline text/src/autogluon/text/automm/predictor.py:614]: We should consider adding support for top_k_average_type=`None` aka just use best checkpoint (equivalent to top_k=1).

File: text/src/autogluon/text/automm/predictor.py (hunk 2 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py:578]: Yes, it's because we need to sort these checkpoints based on their performance on the validation set.

Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py:578]: In fact, I feel that we may use `minmax_mode` or `mode`, which will support "max" and "min". This is the design adopted in packages such as PT Lightning: https://pytorch-lightning.readthedocs.io/en/stable/common/early_stopping.html and Ray Tune: https://docs.ray.io/en/latest/tune/api_docs/suggestion.html . This makes us consistent to the other packages.

File: text/src/autogluon/text/automm/predictor.py [hunk 6]
```diff
@@ -566,21 +566,67 @@ def _fit(
...
+            # Average all the ingredients
             avg_state_dict = average_checkpoints(
-                all_state_dicts=all_state_dicts,
+                all_state_dicts=ingredients,
                 out_path=top_k_avg_ckpt_path,
-                ckpt_template=ckpt_template,
+                ckpt_template=ckpt_template
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:623]: It would be better to wrap this part into a function and put it utils.py. Inside the function, we can have three choices: greedy soup, average, and best. We can do unit tests for them on validation set, i.e., performance of greedy soup is better than average and best.

File: text/src/autogluon/text/automm/predictor.py [hunk 7]
```diff
@@ -566,21 +567,66 @@ def _fit(
...
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:623]: It would be better to add a `top_k_average` function and put it utils.py. Inside the function, we can have three choices: greedy soup, average, and best. We can do unit tests for them on validation set, i.e., performance of greedy soup is better than average and best.

**Comment by szha** [comment]
Job PR-1613-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/4/index.html

**Comment by szha** [comment]
Job PR-1613-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/5/index.html

File: text/src/autogluon/text/automm/predictor.py (hunk 3 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py]: We may do it later. It will internally trigger `self.evaluate` so won't be that easy to wrap in a single function.

File: text/src/autogluon/text/automm/predictor.py (hunk 5 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py]: Revised the design that it will accept "greedy_soup" and "union_soup". For the other keys, it will use the model that has the best validation score.

**Comment by szha** [comment]
Job PR-1613-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/6/index.html

**Comment by szha** [comment]
Job PR-1613-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/10/index.html

File: text/src/autogluon/text/automm/predictor.py (hunk 5 above)
Comment (Innixma) [reply text/src/autogluon/text/automm/predictor.py]: I would insist on adding the exception if it is an invalid key, otherwise someone could typo the key name and have no idea that it isn't working as intended.

**Comment by szha** [comment]
Job PR-1613-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/15/index.html

File: text/src/autogluon/text/automm/configs/optimization/adamw.yaml [hunk 8]
```diff
@@ -14,5 +14,4 @@ optimization:
   val_check_interval: 0.5
   top_k: 3
   top_k_average_method: 'greedy_soup'  # We support averaging method described in https://arxiv.org/pdf/2203.05482.pdf.
-                                       # Currently support "union_soup" and "greedy_soup".
-                                       # For other keys, the best model will be used.
+                                       # Currently support "union_soup", "greedy_soup", "max".
```
Comment (Innixma) [inline text/src/autogluon/text/automm/configs/optimization/adamw.yaml:17]: Why called "best" in error but "max" in doc?

Comment (sxjscience) [reply text/src/autogluon/text/automm/configs/optimization/adamw.yaml:17]: It’s a typo.

File: text/src/autogluon/text/automm/predictor.py [hunk 9]
```diff
@@ -569,21 +569,83 @@ def _fit(
...
+                        )
+                        candidate_performance.append(self.evaluate(val_df,
+                                                                   [validation_metric_name])[validation_metric_name])
+                    all_state_dicts = [v[0] for v in sorted(zip(all_state_dicts, candidate_performance),
+                                                            key=lambda ele: ele[1], reverse=(minmax_mode == 'max'))]
+                if config.optimization.top_k_average_method == 'greedy_soup':
+                    # Select the ingredients based on the methods proposed in paper
+                    #  "Model soups: averaging weights of multiple fine-tuned models improves accuracy without
+                    #  increasing inference time", https://arxiv.org/pdf/2203.05482.pdf
+                    monitor_op = {"min": operator.lt, "max": operator.gt}[minmax_mode]
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:607]: Should they be `operator.le` and `operator.ge`?

File: text/src/autogluon/text/automm/predictor.py [hunk 10]
```diff
@@ -569,21 +569,83 @@ def _fit(
...
+            # TODO(?) Currently, all top-K checkpoints are loaded to the main memory.
+            #  If the checkpoints are really large (e.g., super-giant backbones), this operation can take a lot of
+            #  CPU memory and may potentially trigger memory issue. Consider to optimize that using the
+            #  offload strategy.
             all_state_dicts, ckpt_template = gather_top_k_ckpts(
                 ckpt_dir=save_path,
-                ckpt_paths=checkpoint_callback.best_k_models.keys(),
+                ckpt_paths=best_k_models_path,
             )
+            if config.optimization.top_k_average_method == 'union_soup':
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/automm/predictor.py:585]: Consider registering `union_soup`, `greedy_soup`, and `best` in constants.py and used the imported names.
 `max` and `min` are already defined in constants.py. Better to use their imported names.

Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py]: Done.

File: text/src/autogluon/text/automm/predictor.py (hunk 9 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py]: Done.

File: text/src/autogluon/text/automm/predictor.py (hunk 7 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py:642]: Per offline discussion, it's more complicated to move to a separate function and we may leave it as future work (e.g., when we will need to implement the greedy soup algorithm elsewhere).

File: text/src/autogluon/text/automm/predictor.py (hunk 2 above)
Comment (sxjscience) [reply text/src/autogluon/text/automm/predictor.py]: Also, no bug in the previous version as we are averaging the checkpoints.

**Comment by szha** [comment]
Job PR-1613-19 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/19/index.html

**Comment by sxjscience** [comment]
I conducted experiments on the text+tabular benchmark and noticed that GreedySoup won't provide consistent performance improvement. Thus, I've updated the default for individual cases.

**Comment by szha** [comment]
Job PR-1613-20 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/20/index.html

**Comment by szha** [comment]
Job PR-1613-21 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/21/index.html

**Comment by szha** [comment]
Job PR-1613-22 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/22/index.html

**Comment by szha** [comment]
Job PR-1613-23 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/23/index.html

**Comment by sxjscience** [comment]
@zhiqiangdon You may review again.

File: text/src/autogluon/text/text_prediction/presets.py [hunk 11]
```diff
@@ -12,18 +12,24 @@ def list_text_presets(verbose=False):
         "default": {
             "model.hf_text.checkpoint_name": "google/electra-base-discriminator",
             "optimization.lr_decay": 0.9,
+            "model.hf_text.max_text_len": 256,
```
Comment (zhiqiangdon) [inline text/src/autogluon/text/text_prediction/presets.py:15]: Does the change (512 -> 256) affect performance? 512 works well on our benchmark datasets.

**Comment by szha** [comment]
Job PR-1613-24 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/24/index.html

**Review by zhiqiangdon — APPROVED** [review]
LGTM. Approved.

**Comment by szha** [comment]
Job PR-1613-26 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1613/26/index.html

**Review by Innixma — APPROVED** [review]
LGTM!
